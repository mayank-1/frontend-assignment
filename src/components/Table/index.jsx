import React, { useCallback, useMemo } from "react";
import PropTypes from "prop-types";

const Table = ({ header, data }) => {
  // Memoize headers to prevent unnecessary re-renders
  const memoizedHeader = useMemo(() => header, [header]);

  // Helper function to render a table cell
  const renderCell = useCallback((col, row) => {
    const value = row[col.key];
    if (col.render) {
      return col.render(value, row);
    }
    return value !== null && value !== undefined ? value : "â€”";
  }, []);

  return (
    <table style={{ borderCollapse: "collapse", width: "100%" }}>
      <thead>
        <tr>
          {memoizedHeader.map((col, index) => (
            <th
              key={col.key || index}
              style={{
                border: "1px solid #ddd",
                padding: "8px",
                textAlign: "left",
                backgroundColor: "#f4f4f4",
                fontWeight: "bold",
              }}
            >
              {col.header}
            </th>
          ))}
        </tr>
      </thead>
      <tbody>
        {data.length > 0 ? (
          data.map((row, rowIndex) => (
            <tr key={rowIndex}>
              {memoizedHeader.map((col, colIndex) => (
                <td
                  key={col.key || colIndex}
                  style={{ border: "1px solid #ddd", padding: "8px" }}
                >
                  {renderCell(col, row)}
                </td>
              ))}
            </tr>
          ))
        ) : (
          <tr>
            <td
              colSpan={memoizedHeader.length}
              style={{
                textAlign: "center",
                padding: "16px",
                color: "#888",
              }}
            >
              No data available
            </td>
          </tr>
        )}
      </tbody>
    </table>
  );
};

Table.propTypes = {
  header: PropTypes.arrayOf(
    PropTypes.shape({
      header: PropTypes.string.isRequired,
      key: PropTypes.string.isRequired,
      render: PropTypes.func, // Optional render function
    })
  ).isRequired,
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default Table;
