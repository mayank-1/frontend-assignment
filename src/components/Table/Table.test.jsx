import { render, screen } from "@testing-library/react"; // Import screen!
import Table from "./index";

describe("Table", () => {
  it("renders correctly with basic props (no data)", () => {
    const header = [
      { header: "Column 1", key: "col1" },
      { header: "Column 2", key: "col2" },
    ];

    render(<Table header={header} />);
    expect(
      screen.getByRole("table", {
        "aria-label": "Data table to show amount pledged and percentage funded",
      })
    ).toMatchSnapshot();
    expect(screen.getByText("No data available")).toBeInTheDocument();
  });

  it("renders correctly with data", () => {
    const header = [
      { header: "Column 1", key: "col1" },
      { header: "Column 2", key: "col2" },
    ];
    const data = [
      { col1: "Value 1", col2: "Value 2" },
      { col1: "Value 3", col2: "Value 4" },
    ];

    render(<Table header={header} data={data} />);
    expect(
      screen.getByRole("table", {
        "aria-label": "Data table to show amount pledged and percentage funded",
      })
    ).toMatchSnapshot();
  });

  it('renders "Loading..." while loading', () => {
    const header = [
      { header: "Column 1", key: "col1" },
      { header: "Column 2", key: "col2" },
    ];

    render(<Table header={header} loading />);
    expect(screen.getByText("Loading...")).toBeInTheDocument();
    expect(
      screen.getByRole("table", {
        "aria-label": "Data table to show amount pledged and percentage funded",
      })
    ).toMatchSnapshot();
  });

  it('renders "No data available" for empty data', () => {
    const header = [
      { header: "Column 1", key: "col1" },
      { header: "Column 2", key: "col2" },
    ];

    render(<Table header={header} data={[]} />);
    expect(screen.getByText("No data available")).toBeInTheDocument();
    expect(
      screen.getByRole("table", {
        "aria-label": "Data table to show amount pledged and percentage funded",
      })
    ).toMatchSnapshot();
  });

  it("renders cell content correctly", () => {
    const header = [
      { header: "Column 1", key: "col1" },
      { header: "Column 2", key: "col2" },
    ];
    const data = [{ col1: "Test Value 1", col2: 123 }];

    render(<Table header={header} data={data} />);
    expect(screen.getByText("Test Value 1")).toBeInTheDocument();
    expect(screen.getByText("123")).toBeInTheDocument();
    expect(
      screen.getByRole("table", {
        "aria-label": "Data table to show amount pledged and percentage funded",
      })
    ).toMatchSnapshot();
  });

  it("renders custom cell content using render prop", () => {
    const header = [
      {
        header: "Column 1",
        key: "col1",
        render: (value) => (
          <span data-testid="custom-render">{`Formatted: ${value * 2}`}</span>
        ),
      },
      { header: "Column 2", key: "col2" },
    ];
    const data = [{ col1: 10, col2: "test" }];

    render(<Table header={header} data={data} />);
    expect(screen.getByTestId("custom-render")).toHaveTextContent(
      "Formatted: 20"
    );
    expect(
      screen.getByRole("table", {
        "aria-label": "Data table to show amount pledged and percentage funded",
      })
    ).toMatchSnapshot();
  });
});
