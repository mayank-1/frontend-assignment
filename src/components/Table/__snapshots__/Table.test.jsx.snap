// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > renders "Loading..." while loading 1`] = `
<table
  aria-label="Data table to show amount pledged and percentage funded"
  role="table"
>
  <thead>
    <tr>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 1
      </th>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colspan="2"
        style="text-align: center; padding: 16px; color: rgb(136, 136, 136);"
      >
        Loading...
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > renders "No data available" for empty data 1`] = `
<table
  aria-label="Data table to show amount pledged and percentage funded"
  role="table"
>
  <thead>
    <tr>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 1
      </th>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colspan="2"
        style="text-align: center; padding: 16px; color: rgb(136, 136, 136);"
      >
        No data available
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > renders cell content correctly 1`] = `
<table
  aria-label="Data table to show amount pledged and percentage funded"
  role="table"
>
  <thead>
    <tr>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 1
      </th>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        Test Value 1
      </td>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        123
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > renders correctly with basic props (no data) 1`] = `
<table
  aria-label="Data table to show amount pledged and percentage funded"
  role="table"
>
  <thead>
    <tr>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 1
      </th>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colspan="2"
        style="text-align: center; padding: 16px; color: rgb(136, 136, 136);"
      >
        No data available
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > renders correctly with data 1`] = `
<table
  aria-label="Data table to show amount pledged and percentage funded"
  role="table"
>
  <thead>
    <tr>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 1
      </th>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        Value 1
      </td>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        Value 2
      </td>
    </tr>
    <tr>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        Value 3
      </td>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        Value 4
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table > renders custom cell content using render prop 1`] = `
<table
  aria-label="Data table to show amount pledged and percentage funded"
  role="table"
>
  <thead>
    <tr>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 1
      </th>
      <th
        style="border: 1px solid #ddd; padding: 8px; text-align: left; background-color: rgb(244, 244, 244); font-weight: bold;"
      >
        Column 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        <span
          data-testid="custom-render"
        >
          Formatted: 20
        </span>
      </td>
      <td
        style="border: 1px solid #ddd; padding: 8px;"
      >
        test
      </td>
    </tr>
  </tbody>
</table>
`;
